### PR TITLE
implement quick sort

### PR DESCRIPTION
**Pull request by BaoThan**

### CODE REVIEW

File: src/sort/sort.cpp [hunk 1]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
 
-void quick_sort(std::vector<uint16_t>& arr) {
-    std::sort(arr.begin(), arr.end()); // TODO: remove this and actually implement the algo.
+void quick_sort(std::vector<uint16_t>& arr, 
+                int start_idx,
+                int end_idx,
+                std::function<void(std::vector<uint16_t>,
+                                   std::unordered_map<uint16_t, std::string>)> display) {
+    if (start_idx < end_idx) {
+        // Find the pivot point at the center of the subarray
```
Comment (truongluan303) [inline src/sort/sort.cpp:173]: Not a very informative comment. Anyone can tell this by just looking at the code. So you can remove this comment.

File: src/sort/sort.cpp [hunk 2]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
+        int right_ptr = end_idx;
+
+        std::unordered_map<uint16_t, std::string> color_map{{start_idx, YELLOW}, {end_idx, YELLOW}, {pivot, GREEN}};
+        display(arr, color_map);
+
+        while (left_ptr <= right_ptr) {
+            // Move left pointer to the right until a value smaller than the pivot is found
+            while (arr[left_ptr] < pivot) {
+                left_ptr++;
+                std::unordered_map<uint16_t, std::string> color_map{{start_idx, YELLOW}, {end_idx, YELLOW}, {left_ptr, RED}, {right_ptr, RED}, {pivot, GREEN}};
```
Comment (truongluan303) [inline src/sort/sort.cpp:185]: Line is too long. Something like this is easier to read.
```suggestion
                std::unordered_map<uint16_t, std::string> color_map {
                    {start_idx, YELLOW}, {end_idx, YELLOW}, {left_ptr, RED},
                    {right_ptr, RED}, {pivot, GREEN}};
```

File: src/sort/sort.cpp [hunk 3]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
+                display(arr, color_map);
+            }
+            // Move right pointer to the left until a value greater than the pivot is found
+            while (arr[right_ptr] > pivot) {
+                right_ptr--;
+                std::unordered_map<uint16_t, std::string> color_map{{start_idx, YELLOW}, {end_idx, YELLOW}, {left_ptr, RED}, {right_ptr, RED}, {pivot, GREEN}};
```
Comment (truongluan303) [inline src/sort/sort.cpp:191]: Line too long

File: src/sort/sort.cpp [hunk 4]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
+                int end_idx,
+                std::function<void(std::vector<uint16_t>,
+                                   std::unordered_map<uint16_t, std::string>)> display) {
+    if (start_idx < end_idx) {
+        // Find the pivot point at the center of the subarray
+        uint16_t pivot = arr[(start_idx + end_idx) / 2];
+        int left_ptr = start_idx;
+        int right_ptr = end_idx;
+
+        std::unordered_map<uint16_t, std::string> color_map{{start_idx, YELLOW}, {end_idx, YELLOW}, {pivot, GREEN}};
```
Comment (truongluan303) [inline src/sort/sort.cpp:178]: Does this line exceed 100 cols? If yes then make break it into multiple lines. Also, adding a space between `color_map` and `{` will be a bit more readable.

**Review by truongluan303 — CHANGES_REQUESTED** [review]

**Comment by truongluan303** [comment]
@BaoThan Make sure you try to run both the program and also the test.

File: src/main.cpp [hunk 5]
```diff
@@ -26,15 +26,17 @@ int main() {
...
-        } else if (user_input == "merge_sort") {
+        } else if (user_input == "merge_sort" || user_input == "merge") {
             visualize_merge_sort();
-        } else if (user_input == "selection_sort") {
+        } else if (user_input == "selection_sort" || user_input == "selection") {
             visualize_selection_sort();
-        } else {
+        } else if (user_input == "quick_sort" || user_input == "quick") {
+            visualize_quick_sort();
+        }else {
```
Comment (truongluan303) [inline src/main.cpp:39]: Missing space

File: src/sort/sort.cpp [hunk 6]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
+        display(arr, color_map);
+
+        while (left_ptr <= right_ptr) {
+            // Move left pointer to the right until a value smaller than the pivot is found
+            while (arr[left_ptr] < pivot) {
+                left_ptr++;
```
[the remark this answers is not in the document]
Comment (BaoThan) [reply src/sort/sort.cpp:184]: not work segmentation fault

File: src/sort/sort.cpp [hunk 7]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
+            while (arr[right_ptr] > pivot) {
+                right_ptr--;
+                std::unordered_map<uint16_t, std::string> color_map{{start_idx, YELLOW}, {end_idx, YELLOW}, {left_ptr, RED}, {right_ptr, RED}, {pivot, GREEN}};
+                display(arr, color_map);
+            }
+            // Swap elements at left_ptr and right_ptr if they are out of order
+            if (left_ptr <= right_ptr) {
+                std::swap(arr[left_ptr], arr[right_ptr]);
+                left_ptr++;
+                right_ptr--;
```
[the remark this answers is not in the document]
Comment (BaoThan) [reply src/sort/sort.cpp:198]: not work

File: src/sort/sort.cpp [hunk 8]
```diff
@@ -164,8 +164,53 @@ void merge_sort(std::vector<uint16_t>& arr,
...
+        while (left_ptr <= right_ptr) {
+            // Move left pointer to the right until a value smaller than the pivot is found
+            while (arr[left_ptr] < pivot) {
+                left_ptr++;
+                std::unordered_map<uint16_t, std::string> color_map{{start_idx, YELLOW}, {end_idx, YELLOW}, {left_ptr, RED}, {right_ptr, RED}, {pivot, GREEN}};
+                display(arr, color_map);
+            }
+            // Move right pointer to the left until a value greater than the pivot is found
+            while (arr[right_ptr] > pivot) {
+                right_ptr--;
```
[the remark this answers is not in the document]
Comment (BaoThan) [reply src/sort/sort.cpp:190]: not work

**Review by truongluan303 — APPROVED** [review]

**Pull request merged**